docs/design/staging-environment-proposal.md

```diff
@@ -0,0 +1,26 @@
+# Chronicle Staging Environment
+
+## High Level Diagram
+
+![Chronicle Staging Diagram](assets/chronicle-staging-diagram.png)
+
+## Overview
+
+1. Github Actions will be used to trigger the workflow for the entire deployment process. It will send a deploy queue entry to a `GCP Cloud Tasks` queue, let’s call it the `deploy-queue` with the necessary metadata for the deploy (docker image tag, git branch, etc). Therefore, we will need a docker registry to maintain the necessary images for the staging instance, we will also have to create a system to remove unused images to reduce cost. We will use the `gcloud` cli to push to the docker registry. Essentially, this Github Actions step will occur after every push to the branch, that way a new environment will essentially be created for that branch on every push.
+
+- `Environment Cleaner` is a service that will be running on the cloud to asynchronously detect branch releases by polling the GitHub API. If a current environment for `branch_1` is still running, but ` branch_1` is deleted, this service will handle the cleanup.
+  - If a branch is deleted, a destroy request to a `GCP Cloud Tasks` queue, let’s call it the `destroy-queue` . The `destroy-queue` entry will then be read by a `GCP Cloud Function` that handles the creation of the environment itself.  
+     \* The reasoning behind this multilayer process is that the `GCP Cloud Function` may have easier access to the core credentials via `GCP Secret Manager` required perform all the necessary installations and deployments. If we can easily inject credentials into the `Environment Cleaner` itself, this step is no longer necessary.
+- Create `deployment.yaml` file and template it for each specific branch. This `deployment.yaml` will the the manifest to deploy each chronicle application to a kubernetes cluster.
+
```

Comment:
Based on your recommendations, I say we combine static hosting and our service endpoint into a single container instead of using cloud storage solutions to host the static files.

This will work quite well and has the added benefit of being a single domain name.

The front-end will access the app using this URL

```
https://some-branch.dev.konomi.ai/some-page
```

and the API will be hosted on

```
https://some-branch.dev.konomi.ai/api/some-endpoint
```

I really like this setup to be honest.

![image](https://user-images.githubusercontent.com/11084692/162100590-a78679b3-ccd7-4a20-a94d-e62203937950.png)

Reply:
https://github.com/GoogleCloudPlatform/nginx-ssl-proxy